lib/translator/translation.cc

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "common.h"
+
+namespace translator {
+
+StatusCode Translation::ScsiToNvme(absl::Span<const uint8_t> scsi_cmd,
+                                   const ScsiContext& scsi_context) {
+  this->pipeline_status = StatusCode::kSuccess;
+  // Verify buffer is large enough to contain opcode (one byte)
+  if (scsi_cmd.size() < 1) {
+    DebugLog("Empty SCSI Buffer");
+    this->pipeline_status = StatusCode::kFailure;
+    return StatusCode::kFailure;
```

Comment:
Are there cases where the caller needs to respond differently to different status codes from this function, or should it always unconditionally proceed to execute commands (if any) and finish with `NvmeToScsi`? If the latter, are we actually getting any value from returning an error code here?

It might be useful to consider whether we have both, and can isolate, cases of "the proximate caller is misusing this API" and "the caller is, through no fault of its own, passing on malformed SCSI data it received and needs us to come up with some sort of sensible response to." For the latter case the caller might not need to care whether the SCSI operation is well-formed.

Reply:
In theory the caller never actually has to respond to status codes and the library will still operate as intended. The main reason I was returning status codes was just to offer some insight to library consumers into the status of the SCSI commands the engine is handling.

I think your point about differentiating between API abuse and legitimate malformed data is really interesting. Do you have any insights into how something like this could be accomplished?

Reply:
If the behavior of the caller should be identical regardless of this method's return value, I'd favor omitting the return value entirely for clarity. Normally I'd expect a library interface returning an error to indicate that no further progress can be made.

> Do you have any insights into how something like this could be accomplished?

In this case, I think the structure of the problem lends itself to a very natural solution: bad SCSI commands should result in genuine SCSI error responses and successful (or entirely omitted) library API results, whereas misuse of the API, where possible, should yield an API-level error instead. It's not obvious to me that there's any need to intermix them to begin with.

In general, if you need to intermix different classes of errors, a good way to distinguish them is by exposing various classifier predicates that can be invoked on an error code. When practical, though, it's always much better to use distinct types or control flow to avoid any risk of confusion.

Reply:
This makes sense, I'll update accordingly